modules/pam_motd/pam_motd.c

```diff
@@ -97,9 +97,192 @@ static void try_to_display_directory(pam_handle_t *pamh, const char *dirname)
     }
 }
 
+/*
+ * Get the number of non-overlapping occurrences of substring DELIM in ARG, 
+ * searching left to right.
+ */
+static uint get_num_occurrences(const char *arg, const char *delim)
+{
+    if (arg == NULL || delim == NULL) {
+	return 0;
+    }
+
+    int num_occurrences = 0;
+    int delim_len = strlen(delim);
+
+    const char *arg_ptr = arg;
+    while ((arg_ptr = strstr(arg_ptr, delim)) != NULL) {
+	num_occurrences++;
+	arg_ptr += delim_len;
+    }
+
+    return num_occurrences;
+}
+
+/*
+ * Split a string ARG into an array of NUM_STRS strings or until strtok_r(3)
+ * returns NULL, whichever is lower. Returns newly allocated array of
+ * strings ARG_SPLIT.
+ */
+static char ** split_string(char *arg, int num_strs, const char *delim)
+{
+    if (arg == NULL || delim == NULL) {
+	return NULL;
+    }
+    if (num_strs < 1) {
+	return NULL;
+    }
+
+    char **arg_split = NULL;
+    char *arg_extracted = NULL;
+    int i = 0;
+
+    arg_split = (char **)malloc(sizeof(char *) * num_strs);
+    if (arg_split == NULL) {
+	return NULL;
+    }
+
+    arg_extracted = strtok_r(arg, delim, &arg);
+    while (arg_extracted != NULL && i < num_strs) {
+	arg_split[i++] = arg_extracted;
+	arg_extracted = strtok_r(NULL, delim, &arg);
+    }
+
+    return arg_split;
+}
+
+/*
+ * Join two strings, allocating a new string and inserting a "/" between them.
+ */
+static char * join_dir_strings(const char *a_str, const char *b_str)
+{
+    if (a_str == NULL || b_str == NULL) {
+	return NULL;
+    }
+
+    char *cat_str = NULL;
+    cat_str = (char *)malloc(sizeof(char) * (strlen(a_str) + strlen(b_str) + strlen("/") + 1));
+    if (cat_str == NULL) {
+	return NULL;
+    }
+
+    strcpy(cat_str, a_str);
+    strcpy(cat_str + strlen(a_str), "/");
+    strcpy(cat_str + strlen(a_str) + strlen("/"), b_str);
+
+    return cat_str;
+}
+
+static int compare_strings(const void * a, const void * b) {
+    const char *a_str = *(char **)a;
+    const char *b_str = *(char **)b;
+
+    if (a_str == NULL && b_str == NULL) {
+        return 0;
+    }
+    else if (a_str == NULL) {
+	return -1;
+    }
+    else if (b_str == NULL) {
+	return 1;
+    }
+    else {
+	return strcmp(a_str, b_str);
+    }
+}
+
+static int filter_dirents(const struct dirent *d)
+{
+    return (d->d_type == DT_REG || d->d_type == DT_LNK);
+}
+
+static void try_to_display_directories_with_overrides(pam_handle_t *pamh, char **motd_dir_path_split, int num_motd_dirs)
+{
+    if (pamh == NULL || motd_dir_path_split == NULL) {
+	return;
+    }
+    if (num_motd_dirs < 1) {
+	return;
+    }
+
+    struct dirent ***namelist = NULL;
```

Comment:
Call namelist better - something like dirlists and dirlists_sizes could be more clear.

Reply:
Addressed - sorry `namelist` was hastily chosen from reading the `dirscan` manpage example, should have thought a bit more about it. Changed to `dirscans` and `dirscans_sizes`, holding the output of a `dirscan`. How does this sound? `dirlists` is also good with me